Introduce Virtio-Vsock support

luigix25 opened the pull request:
This PR introduces a wrapper to the virtio-drivers crate like what has been done for virtio-blk. This provides blocking calls to `connect` `close` `send` and `recv`.

Vsock is more flexible than the serial port as it has the concept of port, and we can reuse vsock for anything else.

I had to make 2 changes to the virtio-drivers crate, the idea is to open a PR in that repo. Before proceeding with the PR, I'd like to take some feedback here, maybe more changes are needed.

To compile it you need to enable the feature `vsock` and `virtio-drivers`.

To test it you can use `make test-in-svsm`

Note: If a vsock device is not attached, svsm will not crash.

luigix25 committed:
Revert "virtio: remove vsock support"

Commit c4719d3 dropped driver support for virtio-vsock because it was not being used.

This reverts this commit because we are introducing vsock support.

This revert does not add back the examples/aarch64/main.rs file as it is not needed.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 committed:
kernel/vsock: introduce virtio-vsock support

Add a virtio-vsock wrapper around virtio-drivers. It provides blocking functions
for connect, send, recv, shutdown and force_shutdown.

To use it you need the `vsock` feature.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 committed:
scripts/launch_guest: add vsock to launch_guest

Add an optional `--vsock cid` parameter. This will attach a virtio-vsock
device to the VM.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 committed:
kernel/vsock: add in-svsm tests for VsockStream

Add some tests that perform some basic checks on vsock functionalities:

- double connection
- recv a buffer from the host
- send a buffer to the host

The vsock server is created on the host using ncat.

Also enable the CI to run the vsock tests: use the PR number for vsock
port and CID to avoid conflicts with other runners.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 commented on aproxy/src/main.rs:
```diff
@@ -0,0 +1,83 @@
+        let _ = fs::remove_file(args.unix.clone());
+    }
+
+    if args.vsock_cid != 0 && args.vsock_port != 0 {
```
I am aware that this code is ugly and can be improved. As I am still new to Rust, I am very open to suggestions.

germag replied:
Ok, so my assumption was incorrect `VsockListener::bind()` and `UnixListener::bind()` are not returning the same type. So, you can abstract that with a `Listener` type (similar to what you did for `Transport`)

luigix25 commented on kernel/src/attest.rs:
```diff
@@ -0,0 +1,393 @@
+        let curve = Curve::new(TpmEccCurve::NistP521).map_err(AttestationError::Crypto)?;
+        let ecc = sc_key_generate(&curve).map_err(AttestationError::Crypto)?;
+
+        let transport = {
```
Probably this code can be improved too. Very open to suggestions.

germag replied:
there is not a lot you can do here without using something like `cfg_if!`,

germag replied:
One thing you can do, is to try to keep all the `#cfg`s together, for instance, could this piece of code be part of `Transport`?
for instance, something like:

```rust
impl Transport<'_> {
    #[cfg(feature = "vsock")]
    fn new() -> Self {
        match VsockStream::connect(1234, get_vsock_attest_port(), 2) {
            Ok(value) => Transport::Vsock(value),
            Err(e) => {
                log::info!("{:?} Falling back to the serial", e);
                create_serial_transport()
            }
        }
    }

    #[cfg(not(feature = "vsock"))]
    fn new() -> Self {
        create_serial_transport()
    }
...
}
```

so, here you just call
```rust
let transport = Transport::new();
```

germag replied:
You can do something like this if you want to be fancy :)
```rust
    #[cfg(feature = "vsock")]
    fn new() -> Self {
        VsockStream::connect(1234, get_vsock_attest_port(), 2)
            .map(|value| Transport::Vsock(value))
            .unwrap_or_else(|e| {
                log::info!("{:?} Falling back to the serial", e);
                create_serial_transport()
            })
    }
```

luigix25 replied:
uh I like this approach, very clean!

stefano-garzarella replied:
BTW I think for long term we should introduce a runtime parameter to specify what transport to use and if vsock is specified, we should avoid to fallback to the serial.

stefano-garzarella commented on kernel/src/vsock/virtio_vsock.rs:
```diff
@@ -0,0 +1,156 @@
+
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize, Self::Err> {
+        self.driver.recv(self.remote_cid, self.local_port, self.remote_port, buf)
+        .map_err(|_| SvsmError::Vsock(VsockError::Failed))
+    }
+}
+
+impl Write for VsockStream {
+    type Err = SvsmError;
+
+    fn write(&mut self, buf: &[u8]) -> Result<usize, Self::Err> {
+        self.driver.send(self.remote_cid, self.local_port, self.remote_port, buf)
+        .map_err(|_| SvsmError::Vsock(VsockError::Failed))
```
Why remapping on a generic `VsockError::Failed` ? can we at least have 2 error (e.g. `VsockError::RecvFailed`, etc.?

stefano-garzarella commented on kernel/src/vsock/virtio_vsock.rs:
```diff
@@ -0,0 +1,156 @@
+            .unwrap_or_default()
+            .iter()
+            .find_map(|a| VirtIOVsockDriver::new(PhysAddr::from(*a)).ok())
+            .ok_or(SvsmError::Vsock(VsockError::Failed))?;
```
Same here, this will be impossible to debug if we remap all the errors on a single one.

stefano-garzarella commented on kernel/src/vsock/virtio_vsock.rs:
```diff
@@ -153,4 +153,73 @@ impl Write for VsockStream {
+        let remote_port = 12345;
+
+        let result = device.connect(cid, local_port, remote_port);
+        assert!(!result.is_err(), "Connection to {cid}:{remote_port} failed");
```
please use .unwrap() in all the place where we have a Result.

luigix25 replied:
I ended up using expect and expect_err when possible.

stefano-garzarella commented on scripts/test-in-svsm.sh:
```diff
@@ -49,6 +54,7 @@ TEST_IO_PID=$!
 
 $SCRIPT_DIR/launch_guest.sh --igvm $SCRIPT_DIR/../bin/coconut-test-qemu.igvm \
     --state "$TEST_DIR/svsm_state.raw" \
+    --vsock \
     --unit-tests $TEST_DIR/pipe || true
```
This change is done in a next commit, so if in the future we will bisect and run test on this commit, this will fail to start

stefano-garzarella commented on aproxy/src/main.rs:
```diff
@@ -26,9 +27,18 @@ struct Args {
+    #[clap(long, conflicts_with_all = ["vsock_cid", "vsock_port"], default_value_t = String::from(""))]
     unix: String,
 
+    /// CID for vsock
+    /// FIXME: idea is to accept connections only from this CID
+    #[clap(long, conflicts_with = "unix", default_value_t = 0)]
+    vsock_cid: u32,
+
+    /// Port for vsock
+    #[clap(long, conflicts_with = "unix", default_value_t = 0)]
+    vsock_port: u32,
+
```
Instead of having 2 parameters, what about a single `--vsock <cid>:<port>` ?

stefano-garzarella replied:
wait, the CID in this case is used just as filter right?
should we support both modes:
1. accept connection from any CID
2. accept connection only from a subset of CIDs

germag replied:
`0` is a valid value for a CID

germag replied:
Using "canary values" in rust is discouraged

stefano-garzarella replied:
agree with @germag, yes please use `Option` to check if they are specified or not.

stefano-garzarella replied:
> > wait, the CID in this case is used just as filter right?
> 
> yep!
> 
> I think we need 2 parameters then. I was thinking about something like `--cid 3,4,5,6`. This parameter would be optional: If it's not set, accept from any CID. WDYT?

Yep, but we can eventually implement later

luigix25 replied:
Thanks for all the feedbacks! For now I'll keep it simple and support any CID!

stefano-garzarella commented on scripts/launch_guest.sh:
```diff
@@ -81,6 +82,12 @@ while [[ $# -gt 0 ]]; do
       ;;
+    --vsock)
+      VIRTIO_ENABLE="x-svsm-virtio-mmio=on"
+      VSOCK_DEVICE+="-global virtio-mmio.force-legacy=false "
```
This is duplicated if `--state` is also specified, can we use another variable, like you did with `VIRTIO_ENABLE` ?

luigix25 replied:
True, what about putting everything together in `VIRTIO_ENABLE`?

stefano-garzarella commented on scripts/launch_guest.sh:
```diff
@@ -81,6 +82,12 @@ while [[ $# -gt 0 ]]; do
+    --vsock)
+      VIRTIO_ENABLE="x-svsm-virtio-mmio=on"
+      VSOCK_DEVICE+="-global virtio-mmio.force-legacy=false "
+      VSOCK_DEVICE+="-device vhost-vsock-device,guest-cid=3 "
```
Why `3` is hardcoded?

stefano-garzarella reviewed (commented):
Thanks for this PR, I did a quick review and left some comments.
I'll think a bit more about `Transport` implementation and Vsock, since I'd like to have more abstraction (e.g. be ready to support a new driver for HyperV vsock)

luigix25 commented on kernel/src/vsock/virtio_vsock.rs:
```diff
@@ -153,4 +153,73 @@ impl Write for VsockStream {
+
+        assert!(n_bytes == buffer.len(), "Received less bytes than requested");
+
+        let stringa = core::str::from_utf8(&buffer).unwrap();
+        log::info!("received: {stringa:?}");
```
I can probably drop the log entirely

luigix25 commented:
v2:

- Addressed most @stefano-garzarella's comments.
- Rebased to latest main to pickup #765 otherwise tests won't run

TODO:

- [ ] use just one parameter for aproxy

luigix25 commented:
v3:

- Rebased to latest upstream to pickup latest version of attestation's PR.
- Addressed all `cargo fmt` issues

TODO:

- use just one parameter for aproxy
- comment unsafe blocks
- [X] Fix cargo fmt problems

germag commented on kernel/src/attest.rs:
```diff
@@ -38,11 +44,52 @@ use serde::Serialize;
+}
+
+#[cfg(feature = "vsock")]
+fn get_vsock_attest_port() -> u32 {
+    pub const VSOCK_ATTEST_DEFAULT_PORT: u32 = 1995;
+
+    let cfg = FwCfg::new(SVSM_PLATFORM.get_io_port());
+    match cfg.get_vsock_attest_port() {
+        Ok(port) => port,
+        Err(_) => VSOCK_ATTEST_DEFAULT_PORT,
+    }
+}
```
why `pub const` inside a function (I mean the `pub`) ?, maybe something like:

```rust
fn get_vsock_attest_port() -> u32 {
    const VSOCK_ATTEST_DEFAULT_PORT: u32 = 1995;

    FwCfg::new(SVSM_PLATFORM.get_io_port())
        .get_vsock_attest_port()
        .unwrap_or(VSOCK_ATTEST_DEFAULT_PORT)
}
```

luigix25 replied:
`pub` is probably a leftover. In this way is much cleaner, thanks!

germag commented on kernel/src/attest.rs:
```diff
@@ -82,7 +109,23 @@ impl TryFrom<Tee> for AttestationDriver<'_> {
+            #[cfg(feature = "vsock")]
+            {
+                match VsockStream::connect(1234, get_vsock_attest_port(), 2) {
+                    Ok(value) => Transport::Vsock(value),
+                    Err(e) => {
+                        log::info!("{:?} Falling back to the serial", e);
+                        create_serial_transport()
+                    }
+                }
+            }
+            #[cfg(not(feature = "vsock"))]
+            {
+                create_serial_transport()
```
Another question is, is it correct for creating the transport,  `VsockStream` is connected (i.e., `VsockStream::connect()`) but is different for  `SerialPort` ?

luigix25 replied:
Yes! `SerialPort` from SVSM's perspective is just a pipe where you can read or write, while vsock requires a port and a connection.

FYI: Serial it is exposed as a unix socket by qemu.

germag commented on kernel/src/attest.rs:
```diff
@@ -82,7 +109,23 @@ impl TryFrom<Tee> for AttestationDriver<'_> {
+                match VsockStream::connect(1234, get_vsock_attest_port(), 2) {
+                    Ok(value) => Transport::Vsock(value),
+                    Err(e) => {
+                        log::info!("{:?} Falling back to the serial", e);
```
btw, is really required to log this?, if yes, should it be `info` or better a `warn` (or even `error`)?

luigix25 replied:
IMO having this info can be useful for debugging! Otherwise you would have no clue about vsock not working.

probably warn/error is better! I'm not sure about the one to use, as we have an error on vsock, but this is not fatal as we have a fallback.

luigix25 commented on aproxy/src/main.rs:
```diff
@@ -41,16 +51,31 @@ fn main() -> anyhow::Result<()> {
+    } else {
+        let listener = UnixListener::bind(args.unix).context("unable to bind to UNIX socket")?;
+        for stream in listener.incoming() {
+            match stream {
+                Ok(mut stream) => {
+                    let mut http_client = backend::HttpClient::new(args.url.clone(), args.backend)?;
+                    attest::attest(&mut stream, &mut http_client)?;
+                }
+                Err(_) => {
+                    panic!("error");
+                }
```
should I pickup those changes or put them in a different PR?

stefano-garzarella replied:
if you need to touch that code in any case, it's fine to include in this PR IMO

luigix25 committed:
virtio-drivers/socket: apply clippy lints

Apply lints suggested by clippy.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

n-ramacciotti committed:
github/workflows: enable vhost support in QEMU build and install netcat

In preparation to vsock support, QEMU needs to be configured with vhost
support enabled. Without the flag `--enable-vhost-kernel`, QEMU would fail
to launch with error:
`qemu-system-x86_64: -device vhost-vsock-device,guest-cid=3: 'vhost-vsock-device' is not a valid device model name`

This flag is disabled by default because of `--without-default-features`

Because no-cc is run as non-root, this causes QEMU to fail with the
error: `vhost-vsock: failed to open vhost device: Permission denied"`

To fix this, setup permissions to /dev/vhost-vsock for non-root.

Add the flag to enable support for vsock in QEMU and install netcat

Co-developed-by: Luigi Leonardi <leonardi@redhat.com>
Signed-off-by: Luigi Leonardi <leonardi@redhat.com>
Signed-off-by: Nicola Ramacciotti <niko.ramak@gmail.com>

luigix25 commented:
v5:
- Dropped fw_cfg usage to change the default vsock attestation port.
- Applied all reviewer's suggestions. (Huge thanks to German and Stefano)
- Commented unsafe blocks
- Applied clippy lints
- Rebased on latest main
- Fixed some CI issues.

TODO:
- [ ] Fix VsockStream behavior when hitting EOF

stefano-garzarella commented on kernel/src/attest.rs:
```diff
@@ -39,21 +43,56 @@ use sha2::{Digest, Sha512};
+
+    #[cfg(feature = "vsock")]
+    fn new() -> Self {
+        const VSOCK_ATTEST_DEFAULT_PORT: u32 = 1995;
```
Also the fallback to the serial port should be documented IMO

stefano-garzarella commented on kernel/src/vsock/virtio_vsock.rs:
```diff
@@ -0,0 +1,255 @@
+                port: remote_port,
+            };
+
+            // In case of error return the bytes read so far
```
Why?
Is this following the implementation?

stefano-garzarella reviewed (commented):
I left some comments, I really suggest checking the commit messages, some are not updated with the current implementation. I left some comments for some of them, but I didn't check all.

luigix25 commented on kernel/src/attest.rs:
```diff
@@ -39,21 +43,56 @@ use sha2::{Digest, Sha512};
+    fn new() -> Self {
+        const VSOCK_ATTEST_DEFAULT_PORT: u32 = 1995;
+
+        match VsockStream::connect(1234, VSOCK_ATTEST_DEFAULT_PORT, 2) {
```
I'll go with option 2.

stefano-garzarella commented on kernel/src/virtio/devices.rs:
```diff
@@ -17,13 +18,45 @@ use virtio_drivers::PAGE_SIZE;
+
+    let driver = cfg
+        .get_virtio_mmio_addresses()
+        .expect("No MMIO slots found");
```
Sorry, can you elaborate a bit more about `VsockStream` in this context?

stefano-garzarella replied:
I guess just fail (e.g. -ENODEV in Linux)

stefano-garzarella commented on kernel/src/vsock/virtio_vsock.rs:
```diff
@@ -0,0 +1,314 @@
+
+            dev.update_credit(server_address, local_port)?;
+
+            if received < buffer.len() && first_clean_pos != buffer.len() {
```
okay, I see it now, I thought the `dev.wait_for_event()` was called before. So at this point, let's do the check at the end, but I'd do an early return:

```
if total_received == buffer.len() {
    break;
}

dev.wait_for_event()?;
```

stefano-garzarella commented on kernel/src/vsock/virtio_vsock.rs:
```diff
@@ -0,0 +1,314 @@
+
+            first_clean_pos += received;
+
+            dev.update_credit(server_address, local_port)?;
```
What happens if this fail?

IIUC we discard all the bytes already read, is this what we want?
Same for wait_for_event().

stefano-garzarella commented on kernel/src/vsock/virtio_vsock.rs:
```diff
@@ -0,0 +1,314 @@
+    status: VsockStreamStatus,
+}
+
+impl VsockStream {
```
Should we move this in the vsock/mod.rs ?

I mean, this should not be coupled with the virtio-vsock driver IMO, so if in the future we will need another driver (e.g. hyper-v), we should be able to just add it.

So maybe we should also add a trait like `trait BlockDriver`  that each driver should implement and that you can use in `VsockStream`.

So, in general I'd like to see a bit more abstraction between VsockStream and vsock driver.

luigix25 replied:
makes sense!

stefano-garzarella commented on kernel/src/vsock/mod.rs:
```diff
@@ -0,0 +1,5 @@
+pub mod error;
+#[cfg(feature = "vsock")]
+pub mod virtio_vsock;
```
You already have this in kernel/src/lib.rs (and it's right):
```
#[cfg(feature = "vsock")]
pub mod vsock;
```

So, now I'm also not sure why you need this here.

luigix25 committed:
kernel: introduce VsockTransport trait for driver abstraction

Define a VsockTransport trait to abstract vsock operations and enable
multiple backend implementations.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 committed:
kernel/vsock: introduce VsockStream

Add VsockStream struct that provides a high-level stream-oriented
interface over the vsock driver API.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 commented:
Should have addressed all @stefano-garzarella concerns (**huge** thanks for the detailed review!)
- Rebased on latest main
- Introduced a VsockDriver trait so it's easier to introduce new vsock implementations
- Introduced a `local_port` counter
- Added vsock transport to `ATTESTATION.md` with examples
- Documented the VsockDriver trait
- Introduced a tracking mechanism for MMIO slot status so we don't probe used slots.
- SVSM dosn't panic if compiled with vsock support but a vsock device is not attached
- Minor changes to commit messages
- Some more refactoring suggested by Stefano

luigix25 commented:
- Removed attestation commits (@stefano-garzarella)

osteffenrh commented on kernel/src/vsock/stream.rs:
```diff
@@ -0,0 +1,180 @@
+            return Err(SvsmError::Vsock(VsockError::DriverError));
+        }
+
+        let local_port = VSOCK_DEVICE.get_first_free_port().unwrap();
```
Can we avoid `unwrap()` here and return an error instead?

osteffenrh commented on kernel/src/vsock/stream.rs:
```diff
@@ -178,3 +178,62 @@ impl Drop for VsockStream {
+
+    #[test]
+    #[cfg_attr(not(test_in_svsm), ignore = "Can only be run inside guest")]
+    fn test_virtio_vsock() {
```
a little nit: can we split the test into the 5 cases outlined in the commit message?

luigix25 commented on kernel/src/virtio/devices.rs:
```diff
@@ -57,9 +92,24 @@ impl VirtIOBlkDevice {
         }))
     }
 }
+impl Drop for VirtIOBlkDevice {
```
> I am not so happy having to implement this destructor for all devices separately.
> 
> A more robust approach could be along the lines of having a slot handle type that the drivers consume and hold. This in turn has a drop method that does the proper reset of the slot.

Could you elaborate a little bit more?

germag commented on kernel/src/virtio/devices.rs:
```diff
@@ -9,21 +9,90 @@ use super::hal::*;
+            .compare_exchange(true, false, Ordering::Relaxed, Ordering::Relaxed)
+            .ok()
+            .map(|_| MMIOSlotGuard { addr: self.addr })
+    }
```
This is incorrect, you are using `self.free` as a lock, so to do synchronization, so you cannot use the `Relaxed` ordering.

Atomics are hard, specially for synchronization, I'll suggest to stick to a Mutex (unless you really know what you are doing). Also, why do you want to be lock free?, you are iterating the whole slot list many times, so the performance gain of using an Atomic is minimal, and IIUC the Mutex will not be under contention so it will be quite fast.

germag commented on kernel/src/vsock/stream.rs:
```diff
@@ -0,0 +1,310 @@
+        match VSOCK_DEVICE.recv(self.remote_cid, self.local_port, self.remote_port, buf) {
+            Ok(some) => Ok(some),
+            Err(_e) => Ok(0),
+        }
```
`VSOCK_DEVICE.recv(self.remote_cid, self.local_port, self.remote_port, buf).or(Ok(0))`

luigix25 replied:
thanks!

stefano-garzarella commented:
> I think there could be a race condition on slots probing: As of now two drivers probe one slot at a time, if the slot does not contain the "correct" device, the driver will check the following slot. In order to do this, it will temporarily lock this slot and release it afterward.
> 
> if two drivers probe concurrently, the first driver could lock the slot required by the second one, failing the probing operation.
> 
> Some changes are required.

Maybe we should implement the mmio slot stuff in a separate PR, since it seems a critical point. IMO we should have a single virtio_init() function that parse the mmio slots (and in the future the DTB), initializing the right drivers. Then the users will pick the drivers they need (e.g. vsock socket -> virtio-vsock driver, storage -> virtio-blk driver, etc.)

luigix25 commented on kernel/src/vsock/stream.rs:
```diff
@@ -0,0 +1,310 @@
+    pub fn shutdown(&mut self) -> Result<(), SvsmError> {
+        if VSOCK_DEVICE.try_get_inner().is_err() {
+            return Err(SvsmError::Vsock(VsockError::DriverError));
+        }
```
very good point, thanks! As you and others suggested I am going to change the MMIO Slot tracking, I'll open a different PR for that.

stefano-garzarella commented on kernel/src/vsock/api.rs:
```diff
@@ -0,0 +1,124 @@
+    ///
+    /// * `Ok(u32)` - An available port number
+    /// * `Err(SvsmError)` if no free port could be found
+    fn get_first_free_port(&self) -> Result<u32, SvsmError>;
```
Why this needs to be provided by each single driver?

stefano-garzarella commented on kernel/src/vsock/virtio_vsock.rs:
```diff
@@ -0,0 +1,212 @@
+            .device
+            .locked_do(|dev| dev.connect(server_address, local_port))
+            .map_err(VsockError::from)?;
+
```
Here we should at least put a comment to explain why we are releasing the lock and take it again in the loop.

stefano-garzarella commented on kernel/src/vsock/virtio_vsock.rs:
```diff
@@ -0,0 +1,212 @@
+            let server_address = VsockAddr {
+                cid: remote_cid,
+                port: remote_port,
+            };
```
This can be moved outside the loop.

stefano-garzarella commented on kernel/src/vsock/virtio_vsock.rs:
```diff
@@ -0,0 +1,212 @@
+use virtio_drivers::device::socket::{ConnectionStatus, VsockAddr};
+use virtio_drivers::transport::DeviceType::Socket;
+
+pub struct VirtIOVsockDriver(Box<VirtIOVsockDevice>);
```
IIUC `VirtIOVsockDriver` will be stored into another Box (line 55), is this one really needed?

luigix25 replied:
right, good point. We are doing the same thing in block, I'll open another PR to fix it.

stefano-garzarella commented on scripts/launch_guest.sh:
```diff
@@ -87,6 +89,13 @@ while [[ $# -gt 0 ]]; do
+    --vsock)
+      VIRTIO_ENABLE="x-svsm-virtio-mmio=on"
+      VIRTIO_CONFIG="-global virtio-mmio.force-legacy=false "
+      VSOCK_DEVICE+="-device vhost-vsock-device,guest-cid=$2 "
```
`+=` is not needed

stefano-garzarella commented on .github/workflows/qemu.yml:
```diff
@@ -117,6 +118,9 @@ jobs:
       - name: Install TPM 2.0 Reference Implementation build dependencies
         run: sudo apt install -y build-essential cmake pkg-config
 
+      - name: Install netcat
```
nit: I'd mention that it's used by test-in-svsm for vsock

` Install netcat used by test-in-svsm for vsock tests`

stefano-garzarella commented on virtio-drivers/src/device/socket/connectionmanager.rs:
```diff
@@ -139,6 +155,9 @@ impl<H: Hal, T: Transport, const RX_BUFFER_SIZE: usize>
         let (_, connection) = get_connection(&mut self.connections, destination, src_port)?;
+        if connection.status != ConnectionStatus::Connected {
+            return Err(SocketError::NotConnected.into());
+        }
```
I see, but in `VsockStream::read()` for example we check if it's closed before calling `recv()`, maybe we should remove also these checks in send() and move everything in VsockStream.

stefano-garzarella commented on virtio-drivers/src/device/socket/connectionmanager.rs:
```diff
@@ -288,6 +315,7 @@ impl<H: Hal, T: Transport, const RX_BUFFER_SIZE: usize>
     pub fn shutdown(&mut self, destination: VsockAddr, src_port: u32) -> Result {
         let (_, connection) = get_connection(&mut self.connections, destination, src_port)?;
 
+        connection.status = ConnectionStatus::Closing;
```
> good point, this can be handled easily at `VsockStream` level. WDYT?

Yeah, I agree.

> 
> I know it's very messy because some checks are at driver level, while some others are at Stream level. But I think it's not possible to move all the checks in one place.

But IIUC the original version doesn't keep track of the connection status in the driver, right?

> 
> let's use `send` as an example:
> 
> As of now `send` in `connection_manager.rs` checks for the connection status. What if it didn't? Suppose peer sent X bytes then sent a shutdown. Not all of these bytes where read, so the connection it is still in the list.

Can we move it in another list? kind of closing list?

> 
> `get_connection` would succeed and the driver would send the packet.
> 
> Unfortunately I cannot enforce this check from VsockStream, or at least I cannot think a way of doing it.
> 
> Same goes for `update_credit`.
> 
> If we want to move away from these checks in `connectiomanager`. I think they can be moved in `virtio_vsock.rs`.

Would be nice to handle the connection status just in `VsockStream`, but we should maybe find a way to notify state changes from the driver. That said, we can fix this eventually later if it becomes too complicated.

stefano-garzarella replied:
> > Can we move it in another list? kind of closing list?
> 
> Probably yes, but the driver needs to be (deeply?) modified. So if we want keep changes as little as possible, not sure this is the best approach. It would be the cleanest solution, this is what linux does, right?

I see and I agree on touching the driver as less as possible (for now).
So, maybe we can remove the ConnectionStatus at all, re-use `connection.peer_requested_shutdown` to avoid sending on closing sockets and add just a function to return to SVSM if a connection is accepted/denied from the other peer.

luigix25 commented on kernel/src/vsock/stream.rs:
```diff
@@ -184,3 +184,133 @@ impl Drop for VsockStream {
+
+        start_vsock_server_host();
+
+        let remote_port: u32 = 12345;
```
I added a new parameter to test-in-svsm to select CID and vsock port.

luigix25 committed:
virtio-drivers/connectionmanager: add `is_local_port_used`

Add a function that checks if a local port is in use, either for
listening or for a connection.

This will be used by VsockDriver to allocate local ports when creating
new streams.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 committed:
virtio-drivers/connectionmanager: track connection established state

The connection status is not tracked in the `Connection` struct, so there
is no way to check whether the handshake has completed.

Introduce the `established` field and set it when a `Connected` or
`ConnectionRequest` event is successfully handled. Expose it via
`is_connection_established()`.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 committed:
virtio-drivers/connectionmanager: reject operations after peer shutdown

`send()` and `update_credit()` do not check whether the peer has
requested shutdown, allowing operations on a connection that the peer
considers closed.

Check `peer_requested_shutdown` before proceeding in both methods.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

stefano-garzarella commented on kernel/src/vsock/mod.rs:
```diff
@@ -0,0 +1,69 @@
+                            Some(VSOCK_MIN_PORT)
+                        } else {
+                            Some(port + 1)
+                        }
```
IIRC in Linux we defined `VMADDR_PORT_ANY` for that, can we define here those `well-know` port/cid and use them, instead of hard coding their values ?

00xc commented on kernel/src/vsock/stream.rs:
```diff
@@ -0,0 +1,301 @@
+    /// - The VSOCK device is not available (`VsockError::DriverError`)
+    /// - The connection is already shutdown (`VsockError::SocketShutdown`)
+    /// - The shutdown operation fails
+    pub fn shutdown(&mut self) -> Result<(), SvsmError> {
```
Are there any users that require manually shutting down the stream? If not, then I would suggest moving all of this into Drop. This would make it impossible to use the stream after it is closed. This should also help deduplicate the logic that is already present in Drop and in this method (e.g. `VSOCK_DEVICE.shutdown()`).

stefano-garzarella replied:
If we don't have a real use case right now, we can even remove it IMO and add back when we need.

luigix25 replied:
ok, I will drop it. Thanks for the feedback

00xc commented on kernel/src/vsock/stream.rs:
```diff
@@ -0,0 +1,301 @@
+    ///
+    /// Returns an error if:
+    /// - The VSOCK device is not available (`VsockError::DriverError`)
+    /// - The stream has been shut down (`VsockError::SocketShutdown`)
```
It seems this does not match the code below (`SocketShutdown` is translated into `Ok(0)`)

luigix25 commented on kernel/src/vsock/virtio_vsock.rs:
```diff
@@ -0,0 +1,232 @@
+}
+
+impl VirtIOVsockDriver {
+    pub fn new(slot: MmioSlot) -> Result<Box<Self>, SvsmError> {
```
This was suggested by Stefano (see https://github.com/coconut-svsm/svsm/pull/960#issuecomment-3883713061) to reduce stack usage.

luigix25 commented on kernel/src/error.rs:
```diff
@@ -183,6 +188,13 @@ impl From<BlockDeviceError> for SvsmError {
 }
 
+#[cfg(feature = "vsock")]
+impl From<VsockError> for SvsmError {
```
I'm starting to think we can drop this entirely, vsock is going to be used using `VsockStream` that returns `SvsmError`.

`VsockTransport` (used by `VsockStream`) also returns `SvsmError`. 

I guess can be dropped

luigix25 commented on kernel/src/vsock/api.rs:
```diff
@@ -0,0 +1,110 @@
+
+use crate::error::SvsmError;
+
+pub trait VsockTransport: Sync + Send {
```
I replicated what has been done in `block` for consistency.

stefano-garzarella commented on kernel/src/vsock/mod.rs:
```diff
@@ -0,0 +1,69 @@
+        for _ in 0..MAX_RETRIES {
+            let candidate_port =
+                self.first_free_port
+                    .fetch_update(Ordering::Relaxed, Ordering::Relaxed, |port| {
```
What are your doubts about this?

luigix25 replied:
nothing in particular, I know that atomics are tricky so I highlighted this.

stefano-garzarella replied:
Regarding the use of `fetch_update()` I think it's okay here, since IIUC it ensures that each caller will receive a unique new value (it detects races and re-call the closure multiple times which is okay).

The `Relax` ordering seems also okay, since the port is then used later taking locks in the device. So for now I don't see anything wrong.

stefano-garzarella replied:
Maybe is worth adding a comment with these reasoning (or others you did), will help us in the future (and reviewers now).

stefano-garzarella commented on virtio-drivers/src/device/socket/connectionmanager.rs:
```diff
@@ -105,6 +107,17 @@ impl<H: Hal, T: Transport, const RX_BUFFER_SIZE: usize>
 
+    /// Returns true if a connection has been established, false otherwise
+    pub fn is_connection_established(
+        &mut self,
```
why `mut` here? what we are going to change?

stefano-garzarella replied:
Ah `get_connection()` requires it, maybe we can add a new method, but we can do later.

stefano-garzarella replied:
old comment, see https://github.com/coconut-svsm/svsm/pull/766#discussion_r3115996114

stefano-garzarella commented on kernel/src/vsock/mod.rs:
```diff
@@ -0,0 +1,70 @@
+pub use error::VsockError;
+/// Well-known CID for the host.
+pub const VMADDR_CID_HOST: u32 = 2;
+pub const VMADDR_PORT_ANY: u32 = u32::MAX;
```
Doc missing here

stefano-garzarella commented on virtio-drivers/src/device/socket/connectionmanager.rs:
```diff
@@ -105,6 +107,17 @@ impl<H: Hal, T: Transport, const RX_BUFFER_SIZE: usize>
 
+    /// Returns true if a connection has been established, false otherwise
+    pub fn is_connection_established(
+        &mut self,
```
I don't like having `mut` here since this function is not really changing any state, but I see it comes from `get_connection()`, so maybe we should add `get_connection_ref()`, but we can eventually do this later.

luigix25 replied:
will do in a future PR!

stefano-garzarella commented on virtio-drivers/src/device/socket/connectionmanager.rs:
```diff
@@ -154,6 +154,9 @@ impl<H: Hal, T: Transport, const RX_BUFFER_SIZE: usize>
     pub fn send(&mut self, destination: VsockAddr, src_port: u32, buffer: &[u8]) -> Result {
         let (_, connection) = get_connection(&mut self.connections, destination, src_port)?;
+        if connection.peer_requested_shutdown {
+            return Err(SocketError::NotConnected.into());
```
I see `PeerSocketShutdown` error in VsockError and this seems the right point where we can use it.

stefano-garzarella commented on virtio-drivers/src/device/socket/connectionmanager.rs:
```diff
@@ -283,6 +286,10 @@ impl<H: Hal, T: Transport, const RX_BUFFER_SIZE: usize>
     pub fn update_credit(&mut self, peer: VsockAddr, src_port: u32) -> Result {
         let (_, connection) = get_connection(&mut self.connections, peer, src_port)?;
+        if connection.peer_requested_shutdown {
+            return Err(SocketError::NotConnected.into());
```
Ditto about `PeerSocketShutdown`

stefano-garzarella commented on kernel/src/vsock/virtio_vsock.rs:
```diff
@@ -0,0 +1,236 @@
+            .locked_do(|dev| dev.connect(server_address, local_port))
+            .map_err(VsockError::from)?;
+
+        loop {
```
For the future, we should add a timeout here to break the loop IMO.

luigix25 replied:
right, makes sense. Will do in a separate PR then.

stefano-garzarella commented on kernel/src/vsock/virtio_vsock.rs:
```diff
@@ -0,0 +1,236 @@
+            total_received += received;
+
+            let result = dev.update_credit(server_address, local_port);
+            if result.is_err() || total_received == buffer.len() {
```
Is it okay to ignore issues on updating the credit ?

Maybe yes, but is worth a comment IMO

luigix25 replied:
the idea here is: if `update_credit` fails I still want to provide the user with the already received bytes.
I'll add a comment

stefano-garzarella commented on scripts/test-in-svsm.sh:
```diff
@@ -44,9 +53,6 @@ mkfifo $TEST_DIR/pipe.out
 # Create a raw disk image (512kB in size) for virtio-blk tests containing random data
 dd if=/dev/urandom of="$TEST_DIR/svsm_state.raw" bs=512 count=1024
 
-test_io $TEST_DIR/pipe.in $TEST_DIR/pipe.out &
-TEST_IO_PID=$!
-
```
Why this change?

luigix25 replied:
test_io uses `VSOCK_PORT` and in case of a user provided one, its processing happens later than the `test_io` call

stefano-garzarella replied:
ah, please pass `VSOCK_PORT` as parameter so it's more clear

luigix25 replied:
i still need this change though. `--vsock-port` is processed _after_  `test_io` is called. So I'd end up using always the default value.

stefano-garzarella replied:
sure, I meant that is okay to do this change, plus use the variable instead of accessing global values

stefano-garzarella commented on scripts/test-in-svsm.sh:
```diff
@@ -28,6 +30,13 @@ test_io(){
+            "03")
+              # virtio-vsock in svsm does not handle half duplex connections.
+              echo -n "hello_world" | ncat --no-shutdown -l --vsock -p $VSOCK_PORT &
+              sleep 1
```
So this mean that each new test related to vsock will take at least 1 second ?

luigix25 replied:
unfortunately, yes. Not sure on how we can synchronize the listener and the svsm

stefano-garzarella replied:
Yep, just confirmed running `test-in-svsm`, no please, we should find something else, maybe using `ss`, or since you are already using python, maybe a simple python script to do both. I mean something like this (quickly asked claude to do it, so untested at all):

```
  "03")                                                                                    
    python3 -c "                                                                           
  import socket, sys, struct                                                               
  sock = socket.socket(socket.AF_VSOCK, socket.SOCK_STREAM)                                
  sock.bind((socket.VMADDR_CID_ANY, $VSOCK_PORT))                                          
  sock.listen(1)                                                                           
  # Server is listening — now it's safe to tell the guest
  sys.stdout.buffer.write(struct.pack('I', $VSOCK_PORT))                                   
  sys.stdout.buffer.flush()                                                                
  conn, _ = sock.accept()
  conn.sendall(b'hello_world')                                                             
  conn.shutdown(socket.SHUT_WR)                                                            
  try:
      while conn.recv(1024):                                                               
          pass              
  except:                                                                                  
      pass                                                                                 
  conn.close()
  sock.close()                                                                             
  " > $PIPE_IN              
    ;;
```

stefano-garzarella reviewed (commented):
LGTM, just minor comments.

stefano-garzarella commented on kernel/src/vsock/stream.rs:
```diff
@@ -175,3 +175,127 @@ impl Drop for VsockStream {
+        assert_eq!(n_bytes, buffer.len(), "Received less bytes than requested");
+
+        let string = core::str::from_utf8(&buffer).unwrap();
+        log::info!("received: {string:?}");
```
This will print:
```
[SVSM] received: "hello_world"
```

Do we really need it?

luigix25 replied:
not really, will drop it.

stefano-garzarella commented on scripts/test-in-svsm.sh:
```diff
@@ -28,6 +30,13 @@ test_io(){
               ;;
+            "03")
+              # virtio-vsock in svsm does not handle half duplex connections.
+              echo -n "hello_world" | ncat --no-shutdown -l --vsock -p $VSOCK_PORT &
```
what about redirecting `ncat` stdout to /dev/null, otherwise I see this:

```
[SVSM] test vsock::stream::tests::test_virtio_vsock_read_shutdown ...
[SVSM] test vsock::stream::tests::test_virtio_vsock_write ...
Hello world![SVSM] test vsock::stream::tests::test_virtio_vsock_write_shutdown ...
[SVSM] All tests passed!
```

stefano-garzarella commented on scripts/test-in-svsm.sh:
```diff
@@ -28,6 +30,13 @@ test_io(){
               sha256sum "$TEST_DIR/svsm_state.raw" | cut -f 1 -d ' ' | xxd -p -r > "$PIPE_IN"
               ;;
+            "03")
+              # virtio-vsock in svsm does not handle half duplex connections.
```
please add that this is related to `--no-shutdown` ncat option (not sure, I guess).
if you don't know all `ncat` options, is hard to understand

stefano-garzarella commented on scripts/test-in-svsm.sh:
```diff
@@ -28,6 +30,13 @@ test_io(){
             "02")
               sha256sum "$TEST_DIR/svsm_state.raw" | cut -f 1 -d ' ' | xxd -p -r > "$PIPE_IN"
               ;;
+            "03")
```
Please add a comment before this like we already did for the other cases to describe what 0x03 is supposed to do

luigix25 commented:
Addressed all comments from reviewers:
- Dropped `shutdown` from VsockStream
- Removed `sleep` from the tests, now we are checking if ncat is listening using `ss`
- Added some comments

stefano-garzarella commented on kernel/src/vsock/error.rs:
```diff
@@ -0,0 +1,21 @@
+    /// The peer socket has shutdown.
+    PeerSocketShutdown,
+    /// The local socket has been shutdown.
+    SocketShutdown,
```
Unused now, right?

stefano-garzarella commented on scripts/test-in-svsm.sh:
```diff
@@ -28,6 +31,29 @@ test_io(){
+              fi
+              # write port number as an unsigned int and not as ascii.
+              # this is to avoid ascii to uint conversion in SVSM
+              python3 -c "import sys,struct; sys.stdout.buffer.write(struct.pack('I', $TEST_VSOCK_PORT))" > $PIPE_IN
```
let's use the same approach of the other tests and print hex values here and convert them in SVSM.

Something like this:

```diff
diff --git a/kernel/src/vsock/stream.rs b/kernel/src/vsock/stream.rs
index 0e0bb354..12b9c404 100644
--- a/kernel/src/vsock/stream.rs
+++ b/kernel/src/vsock/stream.rs
@@ -137,7 +137,6 @@ impl Drop for VsockStream {
 #[cfg(all(test, test_in_svsm))]
 mod tests {
     use crate::{testutils::has_test_iorequests, vsock::VMADDR_CID_HOST};
-    use zerocopy::IntoBytes;
 
     use super::*;
 
@@ -145,16 +144,17 @@ mod tests {
         use crate::serial::Terminal;
         use crate::testing::{IORequest, svsm_test_io};
 
-        let mut sp = svsm_test_io().unwrap();
+        let sp = svsm_test_io().unwrap();
 
         sp.put_byte(IORequest::StartVsockServer as u8);
 
-        let mut vsock_port: u32 = 0;
-
-        sp.read(vsock_port.as_mut_bytes())
-            .expect("unable to get the vsock port");
+        // Read port as 4 raw bytes (big-endian, sent by the host via xxd -p -r)
+        let mut port_bytes = [0u8; 4];
+        for byte in &mut port_bytes {
+            *byte = sp.get_byte();
+        }
 
-        vsock_port
+        u32::from_be_bytes(port_bytes)
     }
 
     #[test]
diff --git a/scripts/test-in-svsm.sh b/scripts/test-in-svsm.sh
index 6fbc32d6..993702ca 100755
--- a/scripts/test-in-svsm.sh
+++ b/scripts/test-in-svsm.sh
@@ -50,9 +50,8 @@ test_io(){
                   echo "ncat failed to start listening on vsock port $TEST_VSOCK_PORT"
                   return 1
               fi
-              # write port number as an unsigned int and not as ascii.
-              # this is to avoid ascii to uint conversion in SVSM
-              python3 -c "import sys,struct; sys.stdout.buffer.write(struct.pack('I', $TEST_VSOCK_PORT))" > $PIPE_IN
+              # write port number as a zero-padded 8-char hex string
+              printf '%08x' "$TEST_VSOCK_PORT" | xxd -p -r > "$PIPE_IN"
               ;;
             "")
                 # skip EOF
```

stefano-garzarella commented on scripts/test-in-svsm.sh:
```diff
@@ -28,6 +31,29 @@ test_io(){
             "02")
               sha256sum "$TEST_DIR/svsm_state.raw" | cut -f 1 -d ' ' | xxd -p -r > "$PIPE_IN"
               ;;
+            # 0x03: Virtio-vsock test: send some data to SVSM
```
Let's document better this operation 0x03: it opens a AF_VSOCK socket listening on the host, returns the port, send `hello_world` on the socket. So IMO this should be written here.

Also, should we check here if we receive the right values from SVSM?

Maybe better to have 2 commands or 3, one send data, one receive data (and check), one just listen.

stefano-garzarella replied:
> Also, should we check here if we receive the right values from SVSM?

Let's avoid this for now, I think it will complicate too much current architecture, so we can skip this...

> 
> Maybe better to have 2 commands or 3, one send data, one receive data (and check), one just listen.

... and this.

Please just update the description, following the other commands, describing what it will return to the guest

stefano-garzarella commented on scripts/test-in-svsm.sh:
```diff
@@ -28,6 +31,29 @@ test_io(){
+              done
+              if [ "$NCAT_READY" = false ]; then
+                  echo "ncat failed to start listening on vsock port $TEST_VSOCK_PORT"
+                  return 1
```
no, please avoid to return here (log is fine), tests should continue to run even if we fail vsock ones. I mean for example we want to continue to serve block and attestation tests.

Better to check if it true and send the port.

luigix25 replied:
>Better to check if it true and send the port.

mmh, I think this would cause SVSM to hang waiting for data indefinitely.

stefano-garzarella replied:
True, but `return 1` will do the same, no?

Maybe we should answer with some kind of error (negative port? PORT_ANY ?).

luigix25 replied:
`PORT_ANY` sounds good to me. I'll add a check in svsm.

stefano-garzarella commented on kernel/src/vsock/mod.rs:
```diff
@@ -0,0 +1,71 @@
+pub use error::VsockError;
+/// Well-known CID for the host.
+pub const VMADDR_CID_HOST: u32 = 2;
+/// Well-known CID that means any port number for binding
```
This is not a CID (Context ID), is a port.

luigix25 commented:
Addressed @stefano-garzarella comments:
- Added and fixed a few comments
- Port is now sent to SVSM using hex and not binary
- In case ncat error, return -1 as a port

stefano-garzarella commented on scripts/test-in-svsm.sh:
```diff
@@ -8,10 +8,13 @@
 set -e
 
 SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
+VSOCK_PORT=12345
+VSOCK_CID=10
```
As default I think it's okay, but maybe in the CI pipeline we should specify them when calling this script (e.g. random value, or based on PR number, etc.) to avoid collision with pipelines running on other PRs on the same worker.

luigix25 commented:
- Now using PR number for CID and port within the CI to avoid conflicts
- Added a mechanism to kill ncat processes.

stefano-garzarella commented on .github/workflows/qemu.yml:
```diff
@@ -125,12 +126,25 @@ jobs:
       - name: Run test-in-svsm in QEMU
         run: |
-          QEMU=./citests/qemu/bin/qemu-system-x86_64 scripts/test-qemu-nocc-svsm.sh
+          # Use the pull request number for the port and the cid to avoid conflicts if two (or more)
+          # actions run in parallel. On push-to-main events the PR number is empty, default to 0.
+          # 1024 is the first non reserved port, 10 is an arbitrary number that should be safe enough
+          # to avoid conflicts.
+          PR_NUM=${{ github.event.pull_request.number }}
+          VSOCK_PORT=$(( ${PR_NUM:-0} + 1024 ))
+          VSOCK_CID=$(( ${PR_NUM:-0} + 10 ))
+          QEMU=./citests/qemu/bin/qemu-system-x86_64 scripts/test-qemu-nocc-svsm.sh --vsock-port $VSOCK_PORT --vsock-cid $VSOCK_CID
```
This change should be moved to the next commit.

stefano-garzarella commented on scripts/test-in-svsm.sh:
```diff
@@ -28,6 +31,32 @@ test_io(){
+              if [ "$NCAT_READY" = false ]; then
+                  echo "ncat failed to start listening on vsock port $TEST_VSOCK_PORT"
+                  # use VMADDR_PORT_ANY to signal the guest that an error has occurred
+                  TEST_VSOCK_PORT=$((0xFFFFFFFF))
```
This will override `TEST_VSOCK_PORT` also for next test, so we should move `TEST_VSOCK_PORT=$3` in the while body, so each time is refreshed.

luigix25 commented:
- moved TEST_VSOCK_PORT in the while
- renamed NCAT_PID_FILE to NCAT_PIDS_FILE
